### PR TITLE
feat(net): add shutdown network signal

### LINKED ISSUE #812: Add a signal to gracefully shut down the networking component
### Describe the feature

A signal should be added that prompts the networking component to stop connecting to new peers, drops all pending connections, tells incoming connections that we are shutting down, and sends disconnect reasons to all connected peers.

It should be possible to know when the shutdown is complete. The `NetworkManager` future should resolve.

A proposed API would be to add a `shutdown` function to `NetworkHandle` that returns a future that resolves once the shutdown is complete.

### Additional context

This is currently blocking #807

### PR DESCRIPTION
**Pull request by entropidelic**
closes #812 

Subtasks:
- [x] stop connecting to peers
- [x] drop pending connections
- [x] tell incoming connections that node will shut down
- [x] disconnect all active sessions 
- [x] write tests

### CODE REVIEW

**Review by mattsse — COMMENTED** [review]
looks good so far,

looks like we need something like an `enum NodeState{Active,ShuttingDown}` ?

**Comment by entropidelic** [comment]
any recommendations on how to stop connectig to peers? Was thinking about adding an if clause in the `on_incoming_pending_session` method from the peers manager to check the node state and return an error if it is shutting down. Something like:

```rust
    /// Invoked when a new _incoming_ tcp connection is accepted.
    ///
    /// returns an error if the inbound ip address is on the ban list or
    /// we have reached our limit for max inbound connections
    pub(crate) fn on_incoming_pending_session(
        &mut self,
        addr: IpAddr,
    ) -> Result<(), InboundConnectionError> {
        if self.ban_list.is_banned_ip(&addr) {
            return Err(InboundConnectionError::IpBanned)
        }
        if !self.connection_info.has_in_capacity() {
            return Err(InboundConnectionError::ExceedsLimit(self.connection_info.max_inbound))
        }
        if matches!(node_state, NodeState::ShuttingDown) {
           return Err(InboundConnectionError::ShuttingDown)
        }

        // keep track of new connection
        self.connection_info.inc_in();
        Ok(())
    }
```

wdyt?

**Comment by onbjerg** [comment]
Are we handling this case @mattsse:

1. Downloader requests header/body
2. Networking is shut down
3. We disconnect from all peers

What response will the header/body stage get?

**Comment by mattsse** [comment]
> What response will the header/body stage get?

a channel closed error or will wait forever, but does this even matter since we're shutting down?

**Comment by entropidelic** [comment]
@mattsse any suggestions on how to drop pending connections?

File: crates/net/network/src/swarm.rs [hunk 1]
```diff
@@ -385,3 +405,10 @@ pub(crate) enum SwarmEvent {
     /// Failed to establish a tcp stream to the given address/node
     OutgoingConnectionError { remote_addr: SocketAddr, peer_id: PeerId, error: io::Error },
 }
+
+#[derive(Default)]
+pub(crate) enum NodeConnectionState {
```
Comment (mattsse) [inline crates/net/network/src/swarm.rs:410]: nit: rename to `NetworkConnectionState`

File: crates/net/network/src/swarm.rs [hunk 2]
```diff
@@ -385,3 +405,10 @@ pub(crate) enum SwarmEvent {
     /// Failed to establish a tcp stream to the given address/node
     OutgoingConnectionError { remote_addr: SocketAddr, peer_id: PeerId, error: io::Error },
 }
+
```
Comment (mattsse) [inline crates/net/network/src/swarm.rs:408]: please add some docs.

File: crates/net/network/src/swarm.rs [hunk 3]
```diff
@@ -256,6 +267,15 @@ where
         }
         None
     }
+
+    /// Set node connection state to `ShuttingDown`
+    pub(crate) fn set_shutdown_connection_state(&mut self) {
```
Comment (mattsse) [inline crates/net/network/src/swarm.rs:272]: nit: rename to `on_shutdown_requested`

File: crates/net/network/src/swarm.rs [hunk 4]
```diff
@@ -256,6 +267,15 @@ where
...
+        self.node_connection_state = NodeConnectionState::ShuttingDown;
+    }
+
+    fn is_node_shutting_down(&self) -> bool {
```
Comment (mattsse) [inline crates/net/network/src/swarm.rs:276]:
```suggestion
    fn is_shutting_down(&self) -> bool {
```

File: crates/net/network/src/session/mod.rs [hunk 5]
```diff
@@ -258,6 +258,12 @@ impl SessionManager {
         }
     }
 
+    pub(crate) fn disconnect_all(&self, reason: Option<DisconnectReason>) {
```
Comment (mattsse) [inline crates/net/network/src/session/mod.rs:261]: please add docs

File: crates/net/network/src/manager.rs [hunk 6]
```diff
@@ -510,6 +510,17 @@ where
             NetworkHandleMessage::DisconnectPeer(peer_id, reason) => {
                 self.swarm.sessions_mut().disconnect(peer_id, reason);
             }
+            NetworkHandleMessage::Shutdown => {
+                // Set connection status to `Shutdown`. This will prevent the node to accept
+                // new incoming connections as well as sending connection requests to newly
+                // discovered nodes.
+                self.swarm.set_shutdown_connection_state();
+
+                // drop pending connections
```
Comment (mattsse) [inline crates/net/network/src/manager.rs:519]: we can do this similarly to how we disconnect active sessions by looping over pending session and sending a disconnect message:

we can finally use this:

https://github.com/paradigmxyz/reth/blob/d08aa94f5d7cee743ab05526dd562fcb8e8c3d42/crates/net/network/src/session/handle.rs#L25-L27

unsure if we can send the disconnect reason here though

**Review by mattsse — CHANGES_REQUESTED** [review]
great progress, left some suggestions.

would also appreciate some tests, see existing connect tests.

File: crates/net/network/src/network.rs [hunk 7]
```diff
@@ -203,6 +203,11 @@ impl NetworkHandle {
     pub fn bandwidth_meter(&self) -> &BandwidthMeter {
         &self.inner.bandwidth_meter
     }
+
+    /// Send message to gracefully shutdown node.
+    pub fn shutdown(&self) {
+        self.send_message(NetworkHandleMessage::Shutdown);
```
Comment (mattsse) [inline crates/net/network/src/network.rs:209]: this should be an async function, see comment here https://github.com/paradigmxyz/reth/issues/812#issuecomment-1401049144

File: crates/net/network/src/session/handle.rs [hunk 8]
```diff
@@ -24,11 +21,20 @@ use tokio::{
 #[derive(Debug)]
 pub(crate) struct PendingSessionHandle {
     /// Can be used to tell the session to disconnect the connection/abort the handshake process.
-    pub(crate) _disconnect_tx: oneshot::Sender<()>,
+    pub(crate) disconnect_tx: mpsc::Sender<()>,
```
Comment (mattsse) [inline crates/net/network/src/session/handle.rs:24]: why do we need to make this an mpsc?

what additional messages besides a disconnect need to be sent?

**Review by mattsse — COMMENTED** [review]
great progress, 
left some comments.

File: crates/net/network/src/network.rs (hunk 7 above)
Comment (gakonst) [reply crates/net/network/src/network.rs:209]: what's the comment? not sure why it needs to be async myself

File: crates/net/network/src/session/active.rs [hunk 9]
```diff
@@ -851,4 +851,29 @@ mod tests {
         assert!(calculate_new_timeout(timeout, rtt * 2) > timeout);
         assert!(calculate_new_timeout(timeout, rtt * 2) < timeout * 2);
     }
+
+    // #[tokio::test(flavor = "multi_thread")]
```
Comment (gakonst) [inline crates/net/network/src/session/active.rs:855]: bring this back before merge

File: crates/net/network/src/session/active.rs [hunk 10]
```diff
@@ -638,7 +638,7 @@ mod tests {
         async fn connect_incoming(&mut self, stream: TcpStream) -> ActiveSession {
             let remote_addr = stream.local_addr().unwrap();
             let session_id = self.next_id();
-            let (_disconnect_tx, disconnect_rx) = oneshot::channel();
+            let (disconnect_tx, disconnect_rx) = mpsc::channel(1);
```
Comment (gakonst) [inline crates/net/network/src/session/active.rs:641]: can leave unchanged?

File: crates/net/network/src/session/handle.rs (hunk 8 above)
Comment (gakonst) [reply crates/net/network/src/session/handle.rs:24]: same q

File: crates/net/network/src/session/mod.rs [hunk 11]
```diff
@@ -754,6 +760,9 @@ async fn authenticate(
     )
     .boxed();
 
+    let disconnect_rx = async { disconnect_rx.recv().await };
+    pin_mut!(disconnect_rx);
```
Comment (gakonst) [inline crates/net/network/src/session/mod.rs:764]: why?

File: crates/net/network/src/swarm.rs [hunk 12]
```diff
@@ -256,6 +271,16 @@ where
...
         None
     }
+
+    /// Set network connection state to `ShuttingDown`
+    pub(crate) fn on_shutdown_requested(&mut self) {
+        self.net_connection_state = NetworkConnectionState::ShuttingDown;
+    }
+
+    /// Checks if the node's network connection state is 'ShuttingDown'
+    fn is_shutting_down(&self) -> bool {
```
Comment (gakonst) [inline crates/net/network/src/swarm.rs:281]:
```suggestion
    #[inline]
    fn is_shutting_down(&self) -> bool {
```

File: crates/net/network/src/network.rs (hunk 7 above)
Comment (mattsse) [reply crates/net/network/src/network.rs:209]: it needs to wait until the shutdown was completed (waiting for response)

File: crates/net/network/src/session/handle.rs (hunk 8 above)
Comment (entropidelic) [reply crates/net/network/src/session/handle.rs:24]: I ran into a difficulty using the oneshot channel, let me know if you have any suggestion to workaround this in a cleaner way:
The send method of the oneshot takes ownership of it, which is inside the `PendingSessionHandle`. When calling the `on_handle_message` method of the network manager, a mutable reference of it is received, and this unables the send method to be used on the pending session handle, which is inside the sessions manager.

The mpsc channel does not have this issue, so I tried that for now altough it does not seem correct.

Comment (mattsse) [reply crates/net/network/src/session/handle.rs:24]: I see, 
in this case, simply use an `Option` and ue `if let Some() disconnect_tx.take()`

File: crates/net/network/src/network.rs [hunk 13]
```diff
@@ -205,8 +205,10 @@ impl NetworkHandle {
     }
 
     /// Send message to gracefully shutdown node.
-    pub fn shutdown(&self) {
-        self.send_message(NetworkHandleMessage::Shutdown);
+    pub async fn shutdown(&self) -> Result<(), oneshot::error::RecvError> {
+        let (tx, rx) = oneshot::channel();
+        let _ = self.send_message(NetworkHandleMessage::Shutdown(tx));
```
Comment (mattsse) [inline crates/net/network/src/network.rs:210]: smol clippy lint

File: crates/net/network/src/network.rs [hunk 14]
```diff
@@ -205,8 +205,10 @@ impl NetworkHandle {
     }
 
     /// Send message to gracefully shutdown node.
-    pub fn shutdown(&self) {
-        self.send_message(NetworkHandleMessage::Shutdown);
+    pub async fn shutdown(&self) -> Result<(), oneshot::error::RecvError> {
```
Comment (mattsse) [inline crates/net/network/src/network.rs:208]: this should include a brief explainer what it actually does:
- disconnect all session
- prevent connections

**Review by mattsse — CHANGES_REQUESTED** [review]
great!
I'd appreciate a test for this, similar to the existing `connect` integration tests.
basically, we want to show that:
* session gets disconnected
* connections are rejected

**Comment by entropidelic** [comment]
@mattsse would be adding some tests now. Sorry for the delay, have been a little bit busy this days. Thank you for the suggestions!

**Review by gakonst — APPROVED** [review]

**Pull request merged**